homeassistant/components/squeezebox_player/media_player.py

```diff
@@ -0,0 +1,223 @@
+
+import asyncio
+
+from aioslimproto.client import PlayerState, SlimClient
+from aioslimproto.const import EventType, SlimEvent
+from aioslimproto.server import SlimServer
+
+from homeassistant.components import media_source
+from homeassistant.components.media_player import (
+    BrowseMedia,
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+)
+from homeassistant.components.media_player.browse_media import (
+    async_process_play_media_url,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_IDLE, STATE_OFF, STATE_PAUSED, STATE_PLAYING
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.dt import utcnow
+
+from .const import DEFAULT_NAME, DOMAIN, PLAYER_EVENT
+
+SUPPORTED_FEATURES = (
+    MediaPlayerEntityFeature.PAUSE
+    | MediaPlayerEntityFeature.VOLUME_SET
+    | MediaPlayerEntityFeature.STOP
+    | MediaPlayerEntityFeature.TURN_ON
+    | MediaPlayerEntityFeature.TURN_OFF
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.PLAY_MEDIA
+    | MediaPlayerEntityFeature.VOLUME_MUTE
+    | MediaPlayerEntityFeature.BROWSE_MEDIA
+)
+
+STATE_MAPPING = {
+    PlayerState.IDLE: STATE_IDLE,
+    PlayerState.PLAYING: STATE_PLAYING,
+    PlayerState.PAUSED: STATE_PAUSED,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Squeezebox MediaPlayer(s) from Config Entry."""
+    slimserver: SlimServer = hass.data[DOMAIN]
+    added_ids = set()
+
+    async def async_add_player(player: SlimClient) -> None:
+        """Add MediaPlayerEntity from SlimClient."""
+        # we delay adding the player a small bit because the player name may be received
+        # just a bit after connect. This way we can create a device reg entry with the correct name
+        # the name will either be available within a few milliseconds after connect or not at all
+        # (its an optional data packet)
+        for _ in range(10):
+            if player.player_id not in player.name:
+                break
+            await asyncio.sleep(0.1)
+        async_add_entities([SqueezeboxPlayer(slimserver, player)])
+
+    async def on_slim_event(event: SlimEvent):
+        """Handle player added/connected event."""
+        if event.player_id in added_ids:
+            return
+        added_ids.add(event.player_id)
+        player = slimserver.get_player(event.player_id)
+        await async_add_player(player)
+
+    # register listener for new players
+    config_entry.async_on_unload(
+        slimserver.subscribe(on_slim_event, EventType.PLAYER_CONNECTED)
+    )
+
+    # add all current items in controller
+    for player in slimserver.players:
+        hass.async_create_task(async_add_player(player))
```

Comment:
Why do we create tasks instead of gathering the work and awaiting it?

Reply:
because of the small wait-for-name above